### PR TITLE
add support for chunked content uploads

### PR DESCRIPTION
**Pull request by evgeni**
this allows uploading big files that cannot be uploaded in one go

### CODE REVIEW

**Comment by ehelms** [comment]
ping @SatelliteQE/quality-engineers can someone review this please?

File: nailgun/entities.py [hunk 1]
```diff
@@ -4827,6 +4836,144 @@ def upload_content(self, synchronous=True, **kwargs):
...
 
+    def content_upload_id(self, synchronous=True, **kwargs):
+        """Create an upload request
+
+        :param synchronous: What should happen if the server returns an HTTP
+            202 (accepted) status code? Wait for the task to complete if
+            ``True``. Immediately return the server's response otherwise.
+        :param kwargs: Arguments to pass to requests.
+        :returns: The server's response, with all JSON decoded.
+        :raises: ``requests.exceptions.HTTPError`` If the server responds with
```
Comment (omaciel) [inline nailgun/entities.py]: Don't you need to explicitly raise this exception by examining the response?

File: nailgun/entities.py [hunk 2]
```diff
@@ -4827,6 +4836,144 @@ def upload_content(self, synchronous=True, **kwargs):
...
+        kwargs = kwargs.copy()  # shadow the passed-in kwargs
+        kwargs.update(self._server_config.get_client_kwargs())
+        response = client.post(self.path('content_uploads'), **kwargs)
+        json = _handle_response(response, self._server_config, synchronous)
+        return json
+
+    def content_upload_put(self, upload_id, filepath, **kwargs):
+        """Upload content chunked into a repository
+
+        Expects an upload_id and a path to a file
```
Comment (omaciel) [inline nailgun/entities.py]: Can you add the corresponding docstring tokens for the required arguments and their types?

File: nailgun/entities.py [hunk 3]
```diff
@@ -4827,6 +4836,144 @@ def upload_content(self, synchronous=True, **kwargs):
...
+
+        :param kwargs: Arguments to pass to requests.
+        :returns: The server's response, with all JSON decoded.
```
Comment (omaciel) [inline nailgun/entities.py]: Where is the returned response object?

File: nailgun/entities.py [hunk 4]
```diff
@@ -4827,6 +4836,144 @@ def upload_content(self, synchronous=True, **kwargs):
...
+                        'multipart': True}
+                client.put(content_upload_path, data, **kwargs)
+
+                offset += len(chunk)
+                chunk = contentfile.read(content_chunk_size)
+
+    def content_upload_delete(self, upload_id, **kwargs):
+        """Delete an (incomplete/imported) upload from a repository
+
+        It expects either a list of uploads or upload_ids (but not both).
```
Comment (omaciel) [inline nailgun/entities.py]: Please add docstring tokens for required parameters

File: nailgun/entities.py [hunk 5]
```diff
@@ -4827,6 +4836,144 @@ def upload_content(self, synchronous=True, **kwargs):
...
+
+        :param kwargs: Arguments to pass to requests.
+        :returns: The server's response, with all JSON decoded.
```
Comment (omaciel) [inline nailgun/entities.py]: Explicit return needed?

**Comment by omaciel** [comment]
Was there a reason why the existing mixings for `delete` and `put` were not used? Also, can we have tests?

File: nailgun/entities.py (hunk 1 above)
Comment (rochacbruno) [reply nailgun/entities.py]: @omaciel you are right, it never raises for 4xx or 5xx, `requests` will raise only if address is not found
 `[Errno -5] No address associated with hostname`)

However as all our other methods are not raising I assume we do not want it to raise error but only logging the errors on logger output.

So we need either to remove all those `:raises:` from all methods here or make `nailgun.client` to raise the exceptions or capture explicitly

Comment (rplevka) [reply nailgun/entities.py]: As the guy who usually look into the test results trying to figure out, what went wrong, I vote for raising a proper exception.

**Comment by evgeni** [comment]
You mean `EntityDeleteMixin` and `EntityUpdateMixin`? Those would operate on the `Repository`, not the upload inside the `Repository`. Also the `EntityUpdateMixin` does not support chunking as we need it (but that could be added, I guess).

Tests? Yes, absolutely, will do.

**Comment by evgeni** [comment]
I realized there actually is a `ContentUpload` entity, which should be extended, instead. → WIP

File: nailgun/entities.py (hunk 1 above)
Comment (evgeni) [reply nailgun/entities.py]: My expectation was that `_handle_response` does this for me as it calls `raise_for_status`?

Either way, I don't think this change would belong into this PR, right?

**Comment by evgeni** [comment]
@SatelliteQE/quality-engineers this now has tests and uses the proper entity, please have another look, thanks!

File: nailgun/entities.py (hunk 1 above)
Comment (rochacbruno) [reply nailgun/entities.py]: @evgeni yeah you are right! I missed the `raise_for_status` there https://github.com/SatelliteQE/nailgun/blob/0d186d822941be318bd637285fd0a75d901f097d/nailgun/entities.py#L113 So no change is needed and actually your PR ir right and don't need to raise explicitly here 
c/c @omaciel

**Comment by ehelms** [comment]
It would also be quite helpful that then this gets merged, a new release is made.

**Comment by rochacbruno** [comment]
@ehelms the maintainers are @omaciel @elyezer @sthirugn and @Ichimonji10 according to https://pypi.org/project/nailgun/ so only they have access to upload the release.
(which is something we can delegate to TraviCI to upload new releases when a new tagged commit passes)

**Comment by elyezer** [comment]
I can make a release, just let me know when I should do it.

**Comment by ehelms** [comment]
Now is good :)

**Comment by rochacbruno** [comment]
One more review/merge needed before release

**Comment by ehelms** [comment]
It's an ACK for me (forgot I was a maintainer on the git repo).

**Comment by evgeni** [comment]
Thanks for merging!
Thanks for releasing!

:fireworks:

**Comment by elyezer** [comment]
@evgeni thank **you** for the PR

**Pull request merged**